### PR TITLE
feat(session): add session keeping with sticky load balancing and threaded audit logs

### PR DESCRIPTION
**Pull request by SantiagoDePolonia**
## What

Requests from one client session (one conversation / one agent task) are now recognized and used in two places:

1. **Sticky load balancing** — a load-balanced virtual model routes every request of a session to the target that served it first, keeping provider prompt caches warm and model behavior consistent mid-conversation. On by default; disable per redirect with `session_affinity: false` (YAML/env/admin API/editor checkbox).
2. **Threaded Audit Logs** — the Audit Logs page gets a "Group by session" toggle (on by default): one row per session showing its latest request with a count badge, an expander on the left unfolds the older requests indented with tree connector lines. Every unfolded row keeps full expansion, panes, and the Interactions drawer.

## How a session is identified

First match wins, zero client configuration needed:

1. **Headers** (built-in registry, source-verified per tool): `X-Session-Id` (OpenCode, Kilo Code), `X-Claude-Code-Session-Id` (Claude Code, documented for gateways), `Session-Id`/`Session_id` (Codex CLI, Roo Code), `X-Litellm-Session-Id`, `Helicone-Session-Id`, `Agent-Session-Id` (Goose).
2. **Body fields**: Anthropic `metadata.user_id` (both Claude Code formats parsed), `session_id`, `litellm_session_id`, `prompt_cache_key` (Zed), `/v1/responses` `conversation`.
3. **Auto-detection**: chat/responses requests are grouped by their conversation opening (model + system + tools + messages through the first user turn), so agents that just resend history (Aider, Cline, Continue) still group.

Config: `SESSION_KEEPING_ENABLED` / `SESSION_AUTO_DETECT` / `SESSION_BUILTIN_RULES` / numbered `SESSION_HEADER_<N>` (+`_TRANSFORM`), or the `session:` YAML block. Credential headers are rejected as sources. Non-UUID ids are scoped by user path so weak client ids cannot collide across tenants.

## Provider-specific / operational notes

- Sticky pins live per instance (like rate-limit counters and the realtime call registry): a restart or another replica just re-pins; 6h idle TTL, 10k cap.
- A pinned target that becomes unavailable or rate-saturated is re-picked by the strategy and the session re-pins; the all-saturated first-target 429 fallback never pins.
- `session_id` is a new indexed column on `audit_logs` (idempotent ALTER migration on SQLite/PostgreSQL, index on MongoDB) and a new tri-state `session_affinity` column on `virtual_models`.
- New admin endpoint `GET /admin/audit/sessions` (threads: latest entry + count + span; filters apply to entries before grouping; sessionless entries are singleton threads); `GET /admin/audit/log` gains a `session_id` filter that works without date params.
- Deferred (documented): `/v1/responses` `previous_response_id` chain-following, failover-sweep re-pinning, do-not-pass stripping for session headers.

## Docs

New `docs/features/session-keeping.mdx`, updates to virtual-models docs, `.env.template`, `config.example.yaml`, CLAUDE.md config reference; swagger regenerated.

## Testing

- Unit/integration: detection precedence + both `metadata.user_id` formats + auto-detect stability; sticky semantics (both strategies, re-pin, no-pin-on-saturation, TTL, prune, eviction); audit column roundtrip + sessions listing verified live against SQLite, PostgreSQL, and MongoDB; stream-entry whitelist guard; admin handler + routes; config env/yaml merge. Dashboard: 376 node tests (grouped merge, live folding, gating, flat-mode regression), svelte-check clean, dist rebuilt.
- Live smoke test: header sessions grouped (count 3), auto-detected conversation chained across turns, Claude Code metadata UUID extracted, threaded dashboard UI verified with screenshots (collapsed + expanded with connector lines).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added session keeping to group related requests for sticky virtual-model routing using detected session ids.
  * Added `session_affinity` support for virtual models in the dashboard and admin/API (configurable to return to stateless balancing).
  * Audit logs now store `session_id`, allow exact filtering, and support a threaded “group by session” view via a new `/admin/audit/sessions` endpoint.
* **Bug Fixes**
  * Improved audit retrieval and pagination normalization to avoid empty/incorrect limit edge cases.
* **Documentation**
  * Updated configuration and feature docs for session keeping/session affinity and the new audit session threading behavior.
* **Tests**
  * Added test coverage for session detection, sticky routing, and session-thread audit grouping.

### CODE REVIEW

File: internal/server/http.go [hunk 1]
```diff
@@ -324,6 +326,13 @@ func New(provider core.RoutableProvider, cfg *Config) *Server {
 		e.Use(TaggingCapture(cfg.Tagging))
 	}
 
+	// Session identification runs after snapshot capture (it reads the captured
+	// headers and body) and before audit logging so entries carry the session id
+	// from creation.
+	if cfg != nil && cfg.SessionDetector != nil {
+		e.Use(SessionCapture(cfg.SessionDetector))
+	}
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/server/http.go]: Fixed in b907b355: SessionCapture now runs after authentication, so weak/auto ids are scoped by the effective user path (managed-key bound path included). Audit entries pick the id up in the post-handler re-read; live previews carry it from the terminal events onward.

File: internal/auditlog/reader_sessions_sql.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		if err != nil {
+			return nil, err
+		}
+		sessions = append(sessions, *summary)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating audit session rows: %w", err)
+	}
+	return &SessionListResult{Sessions: sessions, Total: total, Limit: limit, Offset: offset}, nil
+}
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/auditlog/reader_sessions_sql.go:66]: Done in b907b355: the (unreleased) idx_audit_session_id is now composite (session_id, timestamp) on SQL and {session_id:1, timestamp:-1} on MongoDB — it backs the thread-detail fetch and the equality-filtered grouping path. The COALESCE partition key itself can't use a plain index; the date-window timestamp index bounds the scan for the default view.

File: internal/auditlog/session_id_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+		// Filters apply to entries before grouping: only sess-b has a 500.
+		status := 500
+		filtered, err := reader.GetSessions(ctx, LogQueryParams{StatusCode: &status, Limit: 10})
+		if err != nil {
+			t.Fatalf("GetSessions with filter failed: %v", err)
+		}
+		if filtered.Total != 1 || len(filtered.Sessions) != 1 || filtered.Sessions[0].SessionID != "sess-b" {
+			t.Fatalf("filtered result = %+v", filtered)
+		}
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/auditlog/session_id_test.go]: Added in b907b355: both backends now assert GetSessions with a SessionID filter returns exactly the one thread with its count and latest entry.

File: internal/session/detect.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+// scopeSessionID namespaces non-UUID ids by user path. UUIDs are globally
+// unique already and stay raw so operators can correlate them with client-side
+// session identifiers.
+func scopeSessionID(id, userPath string) string {
+	if userPath == "" || uuidRegex.MatchString(id) {
+		return id
+	}
+	return userPath + "|" + id
+}
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/session/detect.go]: Fixed in b907b355 with the suggested hash approach (sha256 over userPath + NUL + id, "scoped-" prefix). Combined with cleanSessionID rejecting control characters, no path/id split can forge another tenant's scoped id.

File: web/dashboard/src/pages/audit-logs/auditList.svelte.js [hunk 5]
```diff
@@ -128,6 +178,70 @@ class AuditListStore {
...
+        },
+      };
+    } catch (e) {
+      console.error("Failed to fetch audit session entries:", e);
+      // Drop the placeholder so the next toggle retries the fetch.
+      const next = { ...liveLogs.auditThreadChildren };
+      delete next[sessionId];
+      liveLogs.auditThreadChildren = next;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply web/dashboard/src/pages/audit-logs/auditList.svelte.js:250]: Fixed in b907b355: the stale branch now drops the loading placeholder (silently, matching stale semantics elsewhere) so the next toggle retries instead of spinning forever.

File: internal/session/detect.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/session/detect.go]: Done in 20759c2d: scopeSessionID now retains sum[:16] (128 bits); the auto-detect id was bumped to match for consistency.

File: internal/auditlog/store_sql.go [hunk 6]
```diff
@@ -108,6 +110,9 @@ var sqlIndexes = []string{
 	"CREATE INDEX IF NOT EXISTS idx_audit_client_ip ON audit_logs(client_ip)",
 	"CREATE INDEX IF NOT EXISTS idx_audit_path ON audit_logs(path)",
 	"CREATE INDEX IF NOT EXISTS idx_audit_user_path ON audit_logs(user_path)",
+	// Composite: serves both the session_id equality filter and its per-thread
+	// timestamp ordering (thread detail and the sessions grouping query).
+	"CREATE INDEX IF NOT EXISTS idx_audit_session_id ON audit_logs(session_id, timestamp)",
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/auditlog/store_sql.go]: Fixed in 20759c2d: the composite index is created as idx_audit_session_timestamp and the migration list drops the old idx_audit_session_id first (same pattern as the existing idx_audit_model drop). Verified against an existing SQLite database that had the single-column index: it is dropped and the composite created on startup. MongoDB index names are key-derived, so no rename was needed there.

File: internal/auditlog/reader_sessions_mongodb_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		bySession, err := reader.GetSessions(ctx, LogQueryParams{SessionID: "sess-a", Limit: 10})
+		if err != nil {
+			t.Fatalf("GetSessions with session filter failed: %v", err)
+		}
+		if bySession.Total != 1 || len(bySession.Sessions) != 1 {
+			t.Fatalf("session-filtered result = %+v", bySession)
+		}
+		if got := bySession.Sessions[0]; got.SessionID != "sess-a" || got.Count != 2 || got.Latest.ID != "a-2" {
+			t.Fatalf("session-filtered thread = %+v", got)
+		}
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/auditlog/reader_sessions_mongodb_test.go]: Done in 20759c2d: both filter scenarios now run as table-driven subtests through a shared assertGetSessionsFilters helper used by the SQL and MongoDB suites, so future filter cases are one table row.

**Comment by SantiagoDePolonia** [comment]
Follow-up review round addressed in 76171be8:

- **[P1] Streamed audits lost session ids** — real regression from moving session detection post-auth: the stream copy is created mid-handler, before the middleware's post-handler enrichment. `CreateStreamEntry` now takes the request context and stamps the id itself, so no call site can miss it (regression test added).
- **[P2] Live rows now re-fold** when a later event delivers the session id — necessary since `audit.started` always precedes detection now; without it, live grouping only converged on refetch.
- **[P2] Whole-session thread expansion** — a `session_id` query without explicit date params was silently getting the default trailing 30-day window; it is now unbounded (explicit dates still apply), matching the documented contract.
- **[P2] Outage-safe pinning** — affinity is keyed to configured targets (`len(entry.targets) > 1`) instead of currently-supported ones, so a session pins its sole viable target during a partial outage and a recovering target can't move an active conversation. The saturated-fallback no-pin rule is unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by SantiagoDePolonia** [comment]
Third review round addressed in fe605c0d:

- **[P1] Atomic affinity assignment** — `stickySessions.resolve` now covers lookup, strategy choice, and pin in one critical section, so concurrent first requests of a session agree on one target (16-goroutine race test added; suite passes under `-race`).
- **[P1] Large/chunked-body detection** — `SessionCapture` runs the shared `requestBodyBytes` materialization for chat/responses requests whose body ingress capture skipped (>64 KiB or unknown length), so body signals and content detection work for large agent conversations; handlers reuse the same buffered read, and bodies beyond the 1 MiB audit bound fall back to header signals (documented).
- **[P2] Recency-aware re-fold** — when completions arrive out of order, the newest request stays thread head (matching what a refresh shows) and the older one becomes the child.
- **Stream-copy identity** — `CreateStreamEntry` now finalizes the full context identity via the existing `EnrichLogEntryWithRequestContext` instead of hand-picking the session id; this also fixes managed-key labels missing from streamed audit records, and there is now exactly one function defining post-auth request metadata used by both streaming and non-streaming paths.

On the architecture suggestions: the atomic-resolve seam matches the proposed `AffinityResolver` shape without introducing the interface yet (per the repo's no-premature-abstraction rule — it's the natural cut point if Redis-backed affinity ever lands); the shared body-materialization stage now runs before detection as suggested. `RequestIdentity`, server-authoritative thread events, and scoping UUID ids (with a separate display value) are noted as follow-ups rather than folded into this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: internal/virtualmodels/balancer.go [hunk 8]
```diff
@@ -50,23 +53,57 @@ func (s *Service) balancedResolution(entry redirectEntry) (core.ModelSelector, b
...
+	affinity := sessionID != "" && entry.sessionAffinity() && len(entry.targets) > 1
+	if affinity {
+		qualified := s.sticky.resolve(entry.vm.Source, sessionID,
+			func(candidate string) bool {
+				_, ok := poolTarget(pool, candidate)
+				return ok
+			},
+			func() string { return pick().qualified },
+			!saturatedFallback,
+		)
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/virtualmodels/balancer.go:100]: Tightened in 47e3d715. Sticky resolution now does a locked fast lookup, performs strategy selection and any catalog reads outside the sticky mutex, then atomically rechecks/installs the pin. Concurrent first requests still converge, and already-pinned requests do not consume round-robin slots (regression coverage added).

File: docs/features/session-keeping.mdx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+3. **Automatic detection** — chat and responses requests with no explicit
+   signal are grouped by their conversation opening: the model, system
+   context, tools, and the messages through the first user turn. Follow-up
+   requests resend that prefix unchanged, so an agent that simply replays its
+   history (Aider, Cline, Continue, …) still gets a stable session id
+   (`auto-…`) with no client changes.
+
+Body-based signals and automatic detection read request bodies up to 1 MiB
+(chunked requests included); larger bodies fall back to header signals.
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply docs/features/session-keeping.mdx:51]: Verified against current code: this suggestion does not apply. The outer server body_size_limit defaults to 10M, but sessionDetectionSnapshot independently limits session inspection to auditlog.MaxBodyCapture (1 MiB), including a limit+1 peek for unknown-length bodies. Bodies above 1 MiB are replayed untouched and fall back to headers, so the existing documentation is accurate.

**Comment by SantiagoDePolonia** [comment]
Final review-comment cleanup in 47e3d715: extracted the shared audit id/request_id matcher noted in the outside-diff review, and moved sticky strategy/catalog selection outside the session mutex while preserving atomic first-pin behavior. Full Go tests, targeted race tests, all dashboard tests, Svelte checks, and commit hooks pass.

File: internal/virtualmodels/balancer.go [hunk 10]
```diff
@@ -50,23 +53,68 @@ func (s *Service) balancedResolution(entry redirectEntry) (core.ModelSelector, b
...
+		choice := pick()
+		qualified := s.sticky.resolve(entry.vm.Source, sessionID,
+			viable,
+			choice.qualified,
+			!saturatedFallback,
+		)
+		if target, ok := poolTarget(pool, qualified); ok {
+			return target.selector, true
+		}
+		return choice.selector, true
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/virtualmodels/balancer.go]: Fixed in 0f9a5f28. The all-saturated honest-429 path now returns before affinity lookup/assignment, so it cannot mutate an existing pin. sticky.lookup also preserves live-but-temporarily-nonviable pins; normal non-saturated re-pinning remains handled atomically by resolve. Added a recovery regression test and verified the virtual-model suite under -race.
